.github/workflows/tests.yml

```diff
@@ -118,20 +107,42 @@ jobs:
         api-level: [ 28, 30 ]
 
     steps:
-      - uses: actions/checkout@v2
+      - uses: actions/checkout@v3
         with:
           submodules: recursive
 
       - name: Set up JDK 11.0.8
-        uses: actions/setup-java@v1
+        uses: actions/setup-java@v3
         with:
+          distribution: 'zulu'
           java-version: 11.0.8
 
+      - uses: gradle/gradle-build-action@v2
+
       - name: Determine emulator target
         id: determine-target
         run: |
           TARGET="google_apis"
           echo "::set-output name=TARGET::$TARGET"
+          
+      - name: AVD cache
+        uses: actions/cache@v3
+        id: avd-cache
+        with:
+          path: |
+            ~/.android/avd/*
+            ~/.android/adb*
+          key: avd-${{ matrix.api-level }}
+
+      - name: Create AVD and generate snapshot for caching
+        if: steps.avd-cache.outputs.cache-hit != 'true'
+        uses: reactivecircus/android-emulator-runner@v2
+        with:
+          api-level: ${{ matrix.api-level }}
+          force-avd-creation: false
+          emulator-options: -no-window -gpu swiftshader_indirect -noaudio -no-boot-anim -camera-back none
+          disable-animations: false
+          script: echo "Generated AVD snapshot for caching."
```

Comment:
https://github.com/ReactiveCircus/android-emulator-runner#usage

Reply:
@Goooler I think you should set arch to x86_64, because Google only provided x86_64 google_apis image for API 29+(I remember it is a breaking version).

Reply:
Good catch! I'll push the commit after this round of tests.

Reply:
39ae2b626a694aada975daa65d3a83be09228858